Objects/setobject.c

```diff
@@ -615,32 +657,75 @@ set_merge(PySetObject *so, PyObject *otherset)
                                  other_entry->hash);
             }
         }
-        return 0;
+        goto exit;
     }
 
     /* We can't assure there are no duplicates, so do normal insertions */
     for (i = 0; i <= other->mask; i++) {
         other_entry = &other->table[i];
         key = other_entry->key;
         if (key != NULL && key != dummy) {
-            if (set_add_entry(so, key, other_entry->hash))
-                return -1;
+            if (set_add_entry(so, key, other_entry->hash)) {
+                rv = -1;
+                goto exit;
+            }
         }
     }
-    return 0;
+
+exit:
+    Py_END_CRITICAL_SECTION2();
+    return rv;
+}
+
+static int set_merge_dict(PySetObject *so, PyObject *other) {
```

Comment:
CPython coding style:

```suggestion
static int
set_merge_dict(PySetObject *so, PyObject *other)
{
```